extensions/amp-form/0.1/test/integration/test-integration-form.js

```diff
@@ -0,0 +1,229 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AmpForm} from '../../amp-form';
+import {timerFor} from '../../../../../src/timer';
+import {AmpMustache} from '../../../../amp-mustache/0.1/amp-mustache';
+import {registerExtendedTemplate,} from
+    '../../../../../src/service/template-impl';
+
+
+describes.realWin('AmpForm Integration', {
+  amp: {
+    runtimeOn: true,
+    ampdoc: 'single',
+  },
+}, env => {
+  const baseUrl = 'http://localhost:31862';
+  let doc;
+  let sandbox;
+  let timer;
+
+  beforeEach(() => {
+    sandbox = env.sandbox;
+    doc = env.win.document;
+    timer = timerFor(env.win);
+    const scriptElement = document.createElement('script');
+    scriptElement.setAttribute('custom-template', 'amp-mustache');
+    doc.body.appendChild(scriptElement);
+    registerExtendedTemplate(env.win, 'amp-mustache', AmpMustache);
+  });
+
+  function getForm(config) {
+    const form = doc.createElement('form');
+    form.id = config.id || 'test-form';
+    form.setAttribute('method', config.method || 'POST');
+    form.setAttribute('target', config.target || '_top');
+
+    const nameInput = doc.createElement('input');
+    nameInput.setAttribute('name', 'name');
+    nameInput.setAttribute('value', 'John Miller');
+    form.appendChild(nameInput);
+    if (config.actionXhr) {
+      form.setAttribute('action-xhr', config.actionXhr);
+    }
+    if (config.action) {
+      form.setAttribute('action', config.action);
+    }
+    const submitButton = doc.createElement('input');
+    submitButton.setAttribute('type', 'submit');
+    form.appendChild(submitButton);
+
+    if (config.on) {
+      form.setAttribute('on', config.on);
+    }
+
+    if (config.successTemplate) {
+      const successDiv = doc.createElement('div');
+      successDiv.setAttribute('submit-success', '');
+      const template = doc.createElement('template');
+      template.setAttribute('type', 'amp-mustache');
+      template./*OK*/innerHTML = config.successTemplate;
+      successDiv.appendChild(template);
+      form.appendChild(successDiv);
+    }
+
+    if (config.errorTemplate) {
+      const successDiv = doc.createElement('div');
+      successDiv.setAttribute('submit-error', '');
+      const template = doc.createElement('template');
+      template.setAttribute('type', 'amp-mustache');
+      template./*OK*/innerHTML = config.errorTemplate;
+      successDiv.appendChild(template);
+      form.appendChild(successDiv);
+    }
+
+    doc.body.appendChild(form);
+    return form;
+  }
+
+  describe('on=submit:form.submit', () => {
+    it('should be protected from recursive-submission', () => {
+      const form = getForm({
+        id: 'sameform',
+        actionXhr: baseUrl + '/form/post',
+        on: 'submit:sameform.submit',
+      });
+      const ampForm = new AmpForm(form);
```

Comment:
I see AmpForm constructor takes in an element and an id? Should we add an id here?